infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/transferprocessor/ParallelTransferProcessor.java

```diff
@@ -0,0 +1,90 @@
+package org.corfudb.infrastructure.log.statetransfer.transferprocessor;
+
+import org.corfudb.infrastructure.log.statetransfer.batch.TransferBatchRequest;
+import org.corfudb.infrastructure.log.statetransfer.batchprocessor.StateTransferBatchProcessor;
+import org.corfudb.infrastructure.log.statetransfer.exceptions.StateTransferBatchProcessorException;
+import org.corfudb.infrastructure.log.statetransfer.exceptions.TransferSegmentException;
+
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Semaphore;
+import java.util.stream.Stream;
+
+import static org.corfudb.infrastructure.log.statetransfer.batch.TransferBatchResponse.TransferStatus.FAILED;
+import static org.corfudb.infrastructure.log.statetransfer.batch.TransferBatchResponse.TransferStatus.SUCCEEDED;
+import static org.corfudb.infrastructure.log.statetransfer.transferprocessor.TransferProcessor.TransferProcessorResult.TransferProcessorStatus.TRANSFER_FAILED;
+import static org.corfudb.infrastructure.log.statetransfer.transferprocessor.TransferProcessor.TransferProcessorResult.TransferProcessorStatus.TRANSFER_SUCCEEDED;
+
+/**
+ * A transfer processor that performs state transfer by allowing only numInFlightRequests at a time.
+ */
+public class ParallelTransferProcessor implements TransferProcessor {
+    /**
+     * A semaphore to control the number of in-flight messages.
+     */
+    private final Semaphore semaphore;
+    /**
+     * A state transfer batch processor that performs the batch transfer.
+     */
+    private final StateTransferBatchProcessor stateTransferBatchProcessor;
+
+    public ParallelTransferProcessor(int numInFlightRequests,
+                                     StateTransferBatchProcessor stateTransferBatchProcessor) {
+        this.semaphore = new Semaphore(numInFlightRequests);
+        this.stateTransferBatchProcessor = stateTransferBatchProcessor;
+    }
+
+    @Override
+    public CompletableFuture<TransferProcessorResult> runStateTransfer
+            (Stream<TransferBatchRequest> batchStream) {
+        // Get a transfer batch request stream iterator
+        Iterator<TransferBatchRequest> iterator = batchStream.iterator();
+
+        // No need to spawn tasks if there is no work to do.
+        if (!iterator.hasNext()) {
+            return CompletableFuture.completedFuture(TransferProcessorResult.builder()
+                    .transferState(TRANSFER_SUCCEEDED).build());
+        }
+
+        CompletableFuture<Void> allFutures = CompletableFuture.completedFuture(null);
+
+        while (iterator.hasNext() && !allFutures.isCompletedExceptionally()) {
```

Comment:
After you fix this, please write a test for this, I think the existing test cases are not covering this.

Reply:
@WenbinZhu  I understand that this check is trying to fast fail to avoid more task being spawned. For those not completed in `allFutures `, it seems like calling `completeExceptionally` does not  cancel tasks behind CF, which means those tasks still consuming resources silently for a while.

According to official doc:

> If not already completed, causes invocations of {@link #get()}
     * and related methods to throw the given exception.

Is it possible to assign an executor when creating tasks by cf? And once we got an exception, we can shutdownNow that executor.

Reply:
We probably need to  change two things: 1) have a variable to record if an exception is thrown in any sub-furture and break out of loop, 2) call `allFutures. completeExceptionally()` after we break out of loop with exception and before we return `allFutures`. 

I'm not sure if your question is on the change above or on the usage of compleExceptionall(), if it's on completeExceptionall(), what I was referring is calling `allFutures. completeExceptionally()`, after this, whoever waits on allFutures will complete immediately with an exception. You can verify this behavior with following code at exits after 2 seconds rather than 10 sec.

```java
    @Test
    public void ttt() {
        CompletableFuture<Void> f = CompletableFuture.runAsync(() -> {
            Sleep.sleepUninterruptibly(Duration.ofSeconds(10));
        });

        Sleep.sleepUninterruptibly(Duration.ofSeconds(3));
        f.completeExceptionally(new WrongEpochException(1));

        f.join();
    }

```

Reply:
@WenbinZhu Yes, my question is about usage of compleExceptionally(). I am saying that call compleExceptionally() does not cancel the real task. In this case, the real task is parallel read and write range.

In below test case, the real task is set flag. And it turns out that even if we call `compleExceptionally()` for all futures, those read and write still occurs..

`@Test
    public void testCompleteExceptionally() {
        AtomicBoolean flag = new AtomicBoolean(false);
        CompletableFuture<Void> f = CompletableFuture.runAsync(() -> {
            Sleep.sleepUninterruptibly(Duration.ofSeconds(10));
            System.out.println("After 10s, set flag to true");
            flag.set(true);
        });

        Sleep.sleepUninterruptibly(Duration.ofSeconds(3));
        f.completeExceptionally(new WrongEpochException(1));

        try {
            f.join();
        } catch (CompletionException e) {
            assertThat(e).hasCauseExactlyInstanceOf(WrongEpochException.class);
        }
        System.out.println("Let's wait 10 more seconds after join");
        Sleep.sleepUninterruptibly(Duration.ofSeconds(10));
        assertThat(flag).isTrue();
    }`

Reply:
@zhangn49 Yea, this can not be entriely avoid, but here what we care about is WrongEpochException, but since we are reading/writing only committed data, it should be fine for now.